sdk/src/main/java/com/google/cloud/dataflow/sdk/coders/Proto2Coder.java

```diff
@@ -0,0 +1,107 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.dataflow.sdk.coders;
+
+import com.google.protobuf.ExtensionRegistry;
+import com.google.protobuf.GeneratedMessage.GeneratedExtension;
+import com.google.protobuf.Message;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+
+/**
+ * An encoder using Google Protocol Buffers 2 binary format.
+ * <p>
+ * To learn more about Protocol Buffers, visit:
+ * https://developers.google.com/protocol-buffers
+ * <p>
+ * To use, specify the {@code Coder} type on a PCollection:
+ * <pre>
+ * {@code
+ * PCollection<MyProtoMessage> records =
+ *     input.apply(...)
+ *          .setCoder(Proto2Coder.of(MyProtoMessage.class));
+ * }
+ * </pre>
+ * <p>
+ * Custom message extensions are also supported, but the coder must be made
+ * aware of them explicitly:
+ * <pre>
+ * {@code
+ * PCollection<MyProtoMessage> records =
+ *     input.apply(...)
+ *          .setCoder(Proto2Coder.of(MyProtoMessage.class)
+ *              .withExtensions(MyProtoMessage.extension1, MyProtoMessage.extension2));
+ * }
+ * </pre>
+ *
+ * @param <T> the type of elements handled by this coder, must extend {@code Message}
+ */
+public class Proto2Coder<T extends Message> extends StandardCoder<T> {
+
+    public static <T extends Message> Proto2Coder<T> of(Class<T> protoMessageClass) {
+      return new Proto2Coder<>(protoMessageClass);
+    }
+
+    private final Class<T> protoMessageClass;
+    private final ExtensionRegistry extensionRegistry = ExtensionRegistry.newInstance();
+
+    protected Proto2Coder(Class<T> protoMessageClass) {
+      this.protoMessageClass = protoMessageClass;
+    }
+
+  /**
+   * Registers Protocol Buffer message extensions with the coder's extension registry.
+   *
+   * @param extensions a variadic list of message extensions the coder uses properly parse messages
+   */
+    public Proto2Coder<T> withExtensions(GeneratedExtension<T, ?>... extensions) {
+      for (GeneratedExtension<T, ?> extension : extensions) {
+        extensionRegistry.add(extension);
+      }
+      return this;
+    }
+
+    @Override
+    public void encode(T value, OutputStream outStream, Context context) throws IOException {
+      value.writeTo(outStream);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T decode(InputStream inStream, Context context) throws IOException {
+      try {
+        T messageInstance = ((T) protoMessageClass.getMethod("getDefaultInstance").invoke(null));
```

Comment:
Instead of using reflection, please store and use the parser for the messages Descriptor.